WordPress/openverse
# Add filtered index and promotion steps to data refresh

## Fixes



Fixes #4149 by @stacimc 

## Description



This PR adds the final steps to the data refresh DAG:
* create and promote the filtered index
* remap indices from the media table to the temp table
* remap constraints from the media table to the temp table
* promote the temp table & delete the original
* promote the 'main' and filtered indices

<img width="1002" alt="Screenshot 2024-08-30 at 4 46 40 PM" src="https://github.com/user-attachments/assets/0301a2dc-74d3-49f5-9103-961b8a09788b">

This was much less straightforward than I imagined and this PR is fairly large as a consequence (although there are tests, and an `es_mapping.py` configuration file that's just copied from the ingestion server adding to the line count). I will attempt to break it into smaller pieces if reviewers like; however I'll point out that it's much easier to manually test in its _current_ state, because you can actually run the entire data refresh (even multiple times in a row).

While working on this PR I also noticed several small issues that I fixed at the same time. I have left comments at the relevant code changes to make these easier to spot:
* I changed the `create_index` steps to create the new index according to a hard-coded index config, rather than basing it off the config of the current index (see comment inline for reasoning)
* I fixed a bug where index calculations were off when multiple data refreshes were run concurrently
* I fixed *several* bugs with error handling and dependency chains, where the failure of a task would not cause presumed downstream tasks to skip. See commit messages
* I fixed a bug in the ingestion server with the way constraints are mapped. See [commit message](https://github.com/WordPress/openverse/pull/4833/commits/4d04e3b9e61897d5a8674b27866c1a914143f8ef)

I've tried to make commit messages and comments in the code as descriptive as possible, but please ping me if clarification is needed anywhere.

## Testing Instructions


Add the following to your `catalog/.env`:
```
AIRFLOW_CONN_SENSITIVE_TERMS="http://catalog_indexer_worker:8003/static/mock_sensitive_terms.txt"
```

Run `ov just recreate` to start with fresh sample data.

Run an image and an audio DAG for a few minutes to get some new records. Wait until you see that a couple 100 lines have been written in the `pull_data` logs, then mark the task as a success. I used Cleveland and Jamendo.

### Establish a baseline

Go to https://app.elasticvue.com/cluster/0/indices and verify you see the four initial indices (`image-init`, `image-init-filtered`, `audio-init`, `audio-init-filtered`), and note the number of records in each. Now run `ov just api/pgcli` and run `describe image;` and `describe audio;`. Save the results somewhere.

### Run the new data refreshes and compare

Now enable the `staging_audio_data_refresh` and `staging_image_data_refresh` DAGs. They should both complete successfully. Once done, check out elasticvue again. You should see that the `-init` indices have all been dropped, and replace with new indices with a uuid (example `audio-f61ab0e757d244ab9abeca0a0613d9e6`, `audio-f61ab0e757d244ab9abeca0a0613d9e6-filtered`, and so on). You should have *more* records in each of these new indices than in the initial ones. To be extra safe, you can go check the logs of the provider DAGs you ran to see how many records were ingested; that's the number of extra records you should see in the non-filtered indices.

In pgcli, again run `describe image;` and `describe audio;`. Compare the results to what you saw earlier; pay particular attention to the indices and constraints. They should be identical, including the index names.



## Checklist


      PRs) or the media properties generator (`ov just catalog/generate-docs media-props`
      for the catalog or `ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
